### ISSUE TITLE
Python 3.4 Windows wheels on PyPI

### ISSUE THREAD

**Issue by mgedmin** [jonparrott/cmarkgfm#2]
Today I noticed that [restview](https://pypi.org/project/restview/) is no longer installable on Windows machines with no C compiler because it indirectly depends on cmarkgfm.  I'm happy to see in the README that you plan to build and upload binary wheels to PyPI.  Do you need help?

(I usually do this sort of thing by cargo-culting [zope.interface's appveyor.yml](https://github.com/zopefoundation/zope.interface/blob/master/appveyor.yml), which is short and simple.  If you've never used Appveyor before, the [password encryption mechanism](https://www.appveyor.com/docs/build-configuration/#secure-variables) may raise questions, but everything else should be self-explanatory.)

**Comment by anthrotype** [comment]
But there are already Windows wheels on PyPI.
https://pypi.org/project/cmarkgfm/#files
What python version are you on?

**Comment by jonparrott** [comment]
Also which pip?

**Comment by hoefling** [comment]
Wheels for Python 3.4 on windows are missing (both the `cp34-cp34m-win32` and `cp34-cp34m-win_amd64`).

**Comment by mgedmin** [comment]
Let me [check](https://jenkins.gedmin.as/job/restview-on-windows/1192/console):

- Python 3.4
- pip 9.0.1

(I didn't notice at first that only one of three tested Python versions failed.  And then I noticed the "Setup automatic wheel building" TODO in the README and made assumptions.)

**Comment by jonparrott** [comment]
We should be able to setup wheels for 3.4 on Windows - we probably just need to add it to the AppVeyor config over at cmarkgfm-wheels.

**Comment by mgedmin** [comment]
I'm not sure if pip installing mingwpy should be sufficient to use it, but it [didn't help](https://ci.appveyor.com/project/jonparrott/cmarkgfm-wheels/build/job/6ws0m6s9089xbd16).  :(

The error AFAICT is because cmark's arena.c uses C99 variable declarations in the middle of the function.  AFAICT the same version of Visual Studio (10.0) is used for Python 3.5, where this works, so the difference must be in compiler flags that distutils uses?

(The upstream project appears to know they're using C99 features not available in older MSVC versions by default, and they have a workaround for that in their build system: [compile in C++ mode](https://github.com/commonmark/cmark/pull/109).  I don't know if that's possible with distutils...)

**Comment by mgedmin** [comment]
Hm, the [What's New in Python 3.5](https://docs.python.org/3/whatsnew/3.5.html) document says

> Windows builds now use Microsoft Visual C++ 14.0, and extension modules should use the same.

so I've misinterpreted [Appveyor's output](https://ci.appveyor.com/project/jonparrott/cmarkgfm-wheels/build/1.0.17/job/u1yvpsw1mhs7f9hs) that mentioned "C:\Program Files (x86)\Microsoft Visual Studio 10.0".

**Comment by anthrotype** [comment]
when using mingwpy, you have to tell distutils/setuptools that you want to use the `mingw32` compiler instead of the default `msvc` one. Try calling bdist_wheel with `--compiler=mingw32`

**Comment by mgedmin** [comment]
How does mingwpy work on Python 2.7, where no such option is passed?

**Comment by anthrotype** [comment]
good question indeed. Try passing -v to `pip wheel`, or call `python setup.py bdist_wheel` which is verbose by default, and confim that gcc is actually been used instead of cl

**Comment by mgedmin** [comment]
Yup, Python 2.7 uses gcc.exe.

I'm having difficulties getting setup.py bdist_wheel to accept a --compiler option.  My latest hack ~~_almost_ works~~ shows some promise, but now cmarkgfm's own build_cmark.py misdetects the compiler and tries to pass `/TP` to `gcc.exe`.   (And as it lives in a separate repository -- nay, the sdist on PyPI actually -- I can't experiment with it in my pull request to the cmarkgfm-wheels repo.)

**Comment by anthrotype** [comment]
> Python 2.7 uses gcc.exe.

I see that, but it's surprising to me. I actually just tried on my Windows VM, installed mingwpy inside a 2.7 virtual environment, then attempted to do python setup.py bdist_wheel from the cmarkgfm git source checkout, without the --compiler=mingw32, and it still attempted to use msvc 9.0 (which I didn't install so it failed). I have to force giving build_ext --compiler=mingw32 to make it use gcc.

and yes, build_cmark.py needs to be told that there can be another compiler type besides "unix" and "msvc", and not pass the extra "/TP" option.

how come it appears to work on python2.7 on Appveyor I wonder..

**Comment by anthrotype** [comment]
oh, it seems that the Appveyor python 2.7 is using the "unix" generated sources -- look at the command line, it mentioned "unix" and doesn't have the extra "/TP" appended:
https://ci.appveyor.com/project/jonparrott/cmarkgfm-wheels/build/1.0.26/job/orxg7bi4nc7xa3o7#L961

this check in build_cmark.py seems to return different things on 2.7 and on 3.4 apparently:
https://github.com/jonparrott/cmarkgfm/blob/f263f81a922b5991a98503b4f90350daf7f97872/src/cmarkgfm/build_cmark.py#L56-L62

on my local VM machine, running that `build_cmark._compiler_type` function returns "mscv" (and the build fails for the same reasons explained by @mgedmin above). Not sure why it would be different for the Appveyor python 2.7, where it's definitely not returning "msvc" for 2.7 (otherwise we would see the extra "/TP" flag, etc.)

**Comment by anthrotype** [comment]
actually the way that _compiler_type function works is by creating a new distutils Distributions and parsing any config files, however it doesn't take into account the command line options passed to setup.py...

the only way I was able to make it work was to modify the `setup.cfg` to add an extra

```
[build]
compiler=mingw32
```

and also patch `build_cmark.py` so that it returns the "unix" sources when compiler is mingw32 (with the windows ones it was failing)

```diff
diff --git a/src/cmarkgfm/build_cmark.py b/src/cmarkgfm/build_cmark.py
index fcd3485..5341c8a 100644
--- a/src/cmarkgfm/build_cmark.py
+++ b/src/cmarkgfm/build_cmark.py
@@ -54,12 +54,14 @@ def _compiler_type():
 
 
 COMPILER_TYPE = _compiler_type()
-if COMPILER_TYPE == 'unix':
+if COMPILER_TYPE in {'unix', 'mingw32'}:^M
     EXTRA_COMPILE_ARGS = ['-std=c99']
     GENERATED_SRC_DIR = UNIX_GENERATED_SRC_DIR
 elif COMPILER_TYPE == 'msvc':
     EXTRA_COMPILE_ARGS = ['/TP']
     GENERATED_SRC_DIR = WIN_GENERATED_SRC_DIR
+else:^M
+    raise AssertionError("unsupported compiler: %s" % COMPILER_TYPE)^M
 
 
 ffibuilder = cffi.FFI()
```

**Comment by anthrotype** [comment]
it also works with a `%HOME%\pydistutils.cfg` file containing the `[build] compiler=mingw32` option, I just tried. We can create that file for appveyor python3.4 (and 2.7 too?) builds

**Comment by hoefling** [comment]
The windows wheels for Python 3.4 are still missing on PyPI.

**Comment by mgedmin** [comment]
AFAIU we're now waiting for the next `cmarkgfm` release, to see what happens to the Appveyor build.

(It might be a good idea to experiment some more in the jonparrott/cmarkgfm-wheels repo, to try building git master rather than the last PyPI release and see if the Python 3.4 build succeeds.  I don't have the energy for that, unfortunately.)

**Comment by jonparrott** [comment]
3.4 failed on AppVeyor; https://ci.appveyor.com/project/jonparrott/cmarkgfm-wheels/build/job/i2leq7h61vtcp8i0

Seems to be still using VS.

You can send a PR to cmarkgfm-wheels to build against master and try to fix 3.4 and we can iterate from there. Or, I can take a look next weekend with a real 3.4 interpreter on my windows box.

**Comment by jonparrott** [comment]
can confirm 3.4 wheels are now building. :) 

https://ci.appveyor.com/project/jonparrott/cmarkgfm-wheels/build/1.0.33

Thanks everyone!
